### ISSUE TITLE
Automatic Seasonal Changes

### ISSUE THREAD

**Issue by aussig** [OpenSceneryX/Library#30]
X-Plane (since 10.41) provides [dataref-driven regions in libraries](https://developer.x-plane.com/2015/03/rfc-dataref-driven-library-regions/), allowing regions to change based on e.g. the seasons.  To date we haven't had any content that could make use of this, but [FlyAgi](https://flyagi.de/viewtopic.php?f=4&t=3) has just integrated TerraMaxx (which exposes its own datarefs for seasons) with her Vegetation library so we could start to do the same. Alternatively, [nicotum-xp](https://forums.x-plane.org/index.php?/files/file/44351-four-seasons/) has written a free plugin `Four Seasons` that seems to do a similar thing to TerraMax, exposing a single dataref for the season.

A quick look and the basic integration seems fairly straightforward.

* [TerraMaxx](http://maxx-xp.com/terramaxx) exposes a set of datarefs such as `maxxxp/seasonsxp/is_summer` and these can be used with `REGION_DREF` statements to use specified named regions when each dataref evaluates to true. 
* [Four Seasons](https://forums.x-plane.org/index.php?/files/file/44351-four-seasons/) exposes a single dataref `nm/four_seasons/season` which contains a numeric value (10 = Spring, 20 = Summer, 30 = Fall, 40 = Winter without ground snow, 45 = Winter with ground snow, 50 = Deep Winter, Only working with Terramax compatible sceneries)
* X-Plane itself exposes `sim/time/local_date_days` which can similarly be used to check the local day in a more granular way. [Einstein's Four Seasons Package](https://forums.x-plane.org/index.php?/files/file/29015-seasons-dataref-driven-libraries/) uses this mechanism.

e.g. for Four Seasons:
```
REGION_DEFINE spring
REGION_RECT -180 -90 179 89
REGION_DREF nm/four_seasons/season == 10

REGION_DEFINE autumn
REGION_RECT -180 -90 179 89
REGION_DREF nm/four_seasons/season == 30

REGION_DEFINE winter
REGION_RECT -180 -90 179 89
REGION_DREF nm/four_seasons/season == 40

REGION_DEFINE winter_deep
REGION_RECT -180 -90 179 89
REGION_DREF nm/four_seasons/season >= 45
```
e.g: for TerraMaxx (doesn't seem to have spring):
```
REGION_DEFINE summer
REGION_RECT -180 -90 179 89
REGION_DREF maxxxp/seasonsxp/is_summer == 1

REGION_DEFINE autumn
REGION_RECT -180 -90 179 89
REGION_DREF maxxxp/seasonsxp/is_autumn == 1

REGION_DEFINE winter
REGION_RECT -180 -90 179 89
REGION_DREF maxxxp/seasonsxp/is_winter == 1

REGION_DEFINE mid_winter
REGION_RECT -180 -90 179 89
REGION_DREF maxxxp/seasonsxp/is_mid_winter == 1
```
e.g. for Native X-Plane, also using a region definition bitmap:
```
REGION_DEFINE northern_hemisphere_spring
REGION_BITMAP northern_hemisphere.png
REGION_DREF sim/time/local_date_days >= 60
REGION_DREF sim/time/local_date_days <= 151

REGION_DEFINE northern_hemisphere_autumn
REGION_BITMAP northern_hemisphere.png
REGION_DREF sim/time/local_date_days >= 244
REGION_DREF sim/time/local_date_days <= 334

REGION_DEFINE northern_hemisphere_winter1
REGION_BITMAP northern_hemisphere.png
REGION_DREF sim/time/local_date_days > 334

REGION_DEFINE northern_hemisphere_winter2
REGION_BITMAP northern_hemisphere.png
REGION_DREF sim/time/local_date_days < 60

REGION_DEFINE southern_hemisphere_spring
REGION_BITMAP southern_hemisphere.png
REGION_DREF sim/time/local_date_days >= 244
REGION_DREF sim/time/local_date_days <= 334

REGION_DEFINE southern_hemisphere_autumn
REGION_BITMAP southern_hemisphere.png
REGION_DREF sim/time/local_date_days >= 60
REGION_DREF sim/time/local_date_days <= 151

REGION_DEFINE southern_hemisphere_winter
REGION_BITMAP southern_hemisphere.png
REGION_DREF sim/time/local_date_days >= 152
REGION_DREF sim/time/local_date_days <= 243
```
Integrating with OpenSceneryX will need researching and planning, but this could be a valuable addition to the library.  Immediate thoughts:

* Does TerraMaxx or Four Seasons give us anything over and above native X-Plane - they change the ground textures too, which is especially important in winter as the trees should match the ground but they may also have better algorithms for determining which season should be applied at the aircraft location.
* If we go with TerraMaxx or Four Seasons, how do we integrate so that users without TerraMaxx or Four Seasons get sensible behaviour
* We should support all available approaches - provide options in installer: No seasons | X-Plane seasons | TerraMaxx | Four Seasons
* Organisation of seasonal items in the library
  * Either: ~~Mark the seasonality in the info.txt file so we can dynamically generate the season regions~~
  * Or: Treat seasonal alternatives as 'hidden' variants of the core object, not exposing them apart from in seasonal REGIONs. This works better as it intrinsically ties the variants to the core object without any additional work.
  * Seasons can then be defined in code and used to generate the appropriate library fragments, which would allow us to support multiple approaches to seasons or to easily adjust things later
  * We cannot have season region name clashes, so it's unsafe to simply dump FlyAgis library.txt files into optional
* There is the thorny issue of snow vs no snow.  Of the two contributions with seasonal variants at time of writing, Einstein's has winter textures with no snow while FlyAgi's has winter texture with snow (both for his winter and deep winter). We need to distinguish so that snowy and non-snowy assets are not mixed, so perhaps we have the following winter seasons:
  * winter_no_snow
  * winter_snow
  * winter_deep_snow
  
  This allows us to deliver the correct asset set to each season region. So for example if TerraMaxx internally shows snowy assets when is_winter == 1 then we inject our snowy set, but if they show non-snowy assets we inject our non-snowy set.

For ref, some useful values for the the `sim/time/local_date_days` datarefs (from Richard Elliott):
```
Jan 1				1
Feb 1				32
Mar 1				60
Apr 1				91
May 1				121
Jun 1				152
Jul 1				182
Aug 1				213
Sep 1				244
Oct 1				274
Nov 1				305
Dec 1				335
 
Spr. Equinox			80
Sum. Solstice		        172
Aut. Equinox			266
Win. Solstice			355
```

**Comment by aussig** [comment]
Propose we don't include seasonal variants as separate items in their own folder, instead we use a suffix to the name and screenshot e.g. `object_winter.obj` and `screenshot_winter.png`

* [x] Library Structure and `info.txt` changes Decision
* [x] Build scripts update to find and copy in seasonal variants
* [x] Build scripts update to support seasonal library paths
* [x] Season `REGION`s must end somehow, before all the standard `EXPORT`s. Perhaps have a summer region that has `REGION_RECT -180 -90 179 89` in its definition, and is inserted at the end.
* [x] Automatically mark items with seasonal variants in `info.txt`

**Comment by marangonico** [comment]
"Four Seasons" author here.

Four Seasons also exposes a Terramax compatible dataref, so going for Terramax support could be a safe option.

What Terramax lacks is Spring support, while Four Seasons implements it (and XP default drefs too).

Keep in mind that you can support multiple scenarios within a single library.txt and without folders duplication. 
Only the library.txt would grows because you have to replicate art assets assignments for every region.

Below is an example of a library.txt that implements seasonal snow and supports Four Seasons and standard drefs.

```
A
800
LIBRARY

PRIVATE

REGION_DEFINE seasonal_snow_four_seasons
REGION_ALL
REGION_DREF nm/four_seasons/season >= 45

REGION_DEFINE North_winter_N
#####     Seasonal region defined by white area of 360x180 bitmap.	Each pixel represents 1x1 degree lat/long
REGION_BITMAP Ahem.png
#####     Winter Jan-Feb in days starting from Jan 1st.
REGION_DREF  sim/time/local_date_days < 90
#####    Assign regional art assets


REGION_DEFINE North_winter_ND
#####     Seasonal region defined by white area of 360x180 bitmap.	Each pixel represents 1x1 degree lat/long
REGION_BITMAP Ahem.png
#####     Winter Dec in days starting from dec 6th.
REGION_DREF  sim/time/local_date_days > 305

REGION seasonal_snow_four_seasons

EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_dry.ter       terrain10/apt_vpol_dry.ter
EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_sdry.ter       terrain10/apt_vpol_sdry.ter
EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_wet.ter       terrain10/apt_vpol_wet.ter
[...]

REGION North_winter_N

EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_dry.ter       terrain10/apt_vpol_dry.ter
EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_sdry.ter       terrain10/apt_vpol_sdry.ter
EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_wet.ter       terrain10/apt_vpol_wet.ter
[...]
REGION North_winter_ND

EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_dry.ter       terrain10/apt_vpol_dry.ter
EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_sdry.ter       terrain10/apt_vpol_sdry.ter
EXPORT_EXCLUDE    lib/g10/terrain10/apt_vpol_wet.ter       terrain10/apt_vpol_wet.ter
```

**Comment by aussig** [comment]
Thank you very much for your input. That almost exactly matches the approach I’ve used - I have `REGION_DEFINE`s and `REGION`s for your plug-in, TerraMaxx and default x-plane and this can be very easily extended to support others (e.g. xAmbience) in future if required. The only annoyance is there is no concept of Boolean ‘or’ in the region defines, otherwise we could support all approaches in a single set of season regions, rather than having duplicated regions.

Please note that [according to this](https://forums.x-plane.org/index.php?/forums/topic/174295-undocumented-region_all-in-librarytxt/&tab=comments#comment-1615956), the command `REGION_ALL` doesn't exist and isn't supported by X-Plane, so I am using `REGION_RECT -180 -90 179 89` instead.

The library build logic is a little more complex than I had hoped, because of snow. Internally, I split winter assets into snow and no snow as I want control of which to use - your plug-in and TerraMaxx create snowy winters so I output snowy assets but core X-Plane ground textures are not snowy so I output non-snow versions if the user chooses that option. 

Would you like to help test all this when it’s ready?

**Comment by marangonico** [comment]
> Would you like to help test all this when it’s ready?

Sure, I'd be very happy to help time permitting.

**Comment by aussig** [comment]
@marangonico In case you missed the announcement on the forums, I've released a beta 4.0.0 version of OpenSceneryX, with all these features - [the testing is being run here](https://forum.thresholdx.net/forum/163-beta-testing-v400/).

**Comment by marangonico** [comment]
Hi,
yes I missed it.. thank you for letting me know.
